sparse/file.go

```diff
@@ -192,46 +194,186 @@ func HashData(data []byte) ([]byte, error) {
 	return sum[:], nil
 }
 
-func GetFiemapExtents(file FileIoProcessor) ([]Extent, error) {
+func GetFileLayout(ctx context.Context, file FileIoProcessor) (<-chan FileInterval, <-chan error, error) {
 	fileInfo, err := file.Stat()
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-	return GetFiemapRegionExts(file, Interval{0, fileInfo.Size()})
+
+	const MaxExtentsBuffer = 1024
+	const MaxInflightIntervals = MaxExtentsBuffer * 10
+	out := make(chan FileInterval, MaxInflightIntervals)
+	errc := make(chan error, 1)
+	go func() {
+		defer close(out)
+		defer close(errc)
+
+		startTime := time.Now()
+		totalExtents := 0
+		defer func() {
+			log.Debugf("retrieved extents for file %v, fileSize: %v elapsed: %.2fs, extents: %v",
+				fileInfo.Name(), fileInfo.Size(),
+				time.Now().Sub(startTime).Seconds(), totalExtents)
+		}()
+
+		var lastIntervalEnd int64
+		isFinalInterval := fileInfo.Size() == 0
+		for !isFinalInterval {
+			exts, err := GetFiemapRegionExts(file, Interval{lastIntervalEnd, fileInfo.Size()}, MaxExtentsBuffer)
+			if err != nil {
+				errc <- err
+				return
+			}
+
+			intervals := make([]FileInterval, 0, len(exts))
+			isFinalInterval = len(exts) == 0
+			totalExtents += len(exts)
+
+			// special case the whole file is a hole
+			if totalExtents == 0 && fileInfo.Size() != 0 {
```

Comment:
If the filesize is 0, there is nothing to be done.
This is just an extra guard I kept, but it's not needed :)
```
		isFinalInterval := fileInfo.Size() == 0
		for !isFinalInterval {
```

Reply:
Oh, this is just a question. I am curious why fileInfo.Size() is not 0 when the whole file is a hole.

Reply:
Thanks for clarifying, I understand the question now.
Let me give an example + explanation:

```shell
$ truncate --size=10K test.truncate
$ ls -lsah test.truncate 
0 -rw-rw-r-- 1 jmoody jmoody 10K May 26 15:00 test.truncate

$ fiemap test.truncate 
File test.truncate has 0 extents:
#	Logical          Physical         Length           Flags
```

From the above example we can see that a file that only consists of a hole, consumes 0 actual space, while the fileinfo.Size() would return 10K as that is specified size for that file.

Since the initial file gets created during the serverOpen call on the remote end, which does a truncate to the specified fileSize. In the 0 size case there is no need to do anything, since the remote file will already be created with 0 size.

Reply:
OK. `fileinfo.Size() ` returns apparent size rather than actual size. I was thinking it returned the actual size, lol.